README.md
# DataTag Helper

[![Open in Visual Studio Code](https://open.vscode.dev/badges/open-in-vscode.svg)](https://open.vscode.dev/allenyllee/DataTag_helper)

[![](assets/interrogate_badge.svg)](https://github.com/econchick/interrogate)

_________________

[Read Latest Documentation](https://allenyllee.github.io/DataTag_helper/) - [Browse GitHub Code Repository](https://github.com/allenyllee/DataTag_helper/)

_________________
![Screenshot-of-DataTag_helper](./assets/Snipaste_2020-06-01_19-59-29.png)


## How to Build

### Method 1: Use Remote Container

#### Step 1: Clone into container

If you have installed [VSCode](https://code.visualstudio.com/) and [Docker](https://www.docker.com/get-started), just click [![Open in Visual Studio Code](https://open.vscode.dev/badges/open-in-vscode.svg)](https://open.vscode.dev/allenyllee/DataTag_helper) and choose "Clone repo in container volume", it'll open a new container with building environments.

#### Step 2: Build

Building linux executable:
```
pyinstaller -F --windowed build-win.spec
```

Building Windows executable:
```
wine pyinstaller -F --windowed build-win.spec
```

#### Step 3: Generate API docs

Hosing in local
```
portray server
```

Push to github pages
```
portray on_github_pages
```


> We use [portary](https://github.com/timothycrosley/portray) to generate docs, and you can add [Pymdown Extensions](https://facelessuser.github.io/pymdown-extensions/) to extend Markdown format. For example, you can add `"pymdownx.arithmatex"` into `extra_markdown_extensions` list which under the section `[tool.portray]` of `pyproject.toml` to have the ability to parse LaTeX math equations.

#### Step 4: test pre-commit hook before commit

test for a file:

```sh
pre-commit run --files ./DataTag_helper.py
```

test for all files:

```sh
pre-commit run --all-files
```

If you want to commit without pre-commit checks, just run

```sh
git commit -m "some messages" --no-verify
```

### Method 2: Manually Setup Environments

### Step 1: Install Anaconda

https://www.anaconda.com/products/individual


### Step 2: Install dependencies

Navigating to the directory of this repo.

```
cd /workspace/DataTag_helper
```

Create conda env

```shell
conda create --name DataTag_helper python=3.6

conda activate DataTag_helper
```
> If conda install has permission error, excute:
>
> ```
> sudo chown 1000:1000 -R /opt/conda/
> ```
>
> This will change the owner of conda folder.

> If conda activation failed, just execute:
> ```
> conda init bash
> ```
> to init bash env

Install packages

```shell
conda install wxpython
conda install -c conda-forge gooey

pip install -r project_requirements.txt
pip install -r requirements.txt
```

> must first install wxpython before Gooey to avoid strange error
>
> [Installation via pip fails needing pathlib2 · Issue #474 · chriskiehl/Gooey](https://github.com/chriskiehl/Gooey/issues/474)




### Step 3: Execute

```shell
python DataTag_helper.py
```

you should see the GUI window popup.

### Step 4: Install pyinstaller

[Gooey/Packaging-Gooey.md at master · chriskiehl/Gooey](https://github.com/chriskiehl/Gooey/blob/master/docs/packaging/Packaging-Gooey.md)

Packing Gooey into a standalone executable is super straight forward thanks to PyInstaller. It is the only dependency you'll need and can be installed via the following.

```shell
pip install pyinstaller==3.5
```

### Step 5: Building linux excutable with .spec file

From the command line, run

```shell
pyinstaller -F --windowed build-win.spec
```

- `-F` tells PyInstaller to create a single bundled output file
- `--windowed` disables the terminal which would otherwise launch when you opened your app.

And that's it. Inside of the `dist/` directory, you'll find a beautiful stand-alone executable that you can distribute to your users.

> If you encounter an error:
>
> ```
> FileNotFoundError: [Errno 2] No such file or directory: 'objcopy': 'objcopy'
> ```
>
> you should install dependenies
>
> ```
> sudo apt update
> sudo apt install -y binutils libc6
> ```
> see: [FileNotFoundError: [Errno 2] No such file or directory: 'objcopy': 'objcopy' · Issue #3815 · pyinstaller/pyinstaller](https://github.com/pyinstaller/pyinstaller/issues/3815)


### Step 6: Install wine

1. add wine apt repository

    For [Ubuntu - WineHQ Wiki](https://wiki.winehq.org/Ubuntu):

    ```shell
    sudo dpkg --add-architecture i386

    wget -O - https://dl.winehq.org/wine-builds/winehq.key | sudo apt-key add -

    sudo add-apt-repository 'deb https://dl.winehq.org/wine-builds/ubuntu/ bionic main'
    ```

    > Install unmet dependence: `faudio` (only needed for unbuntu18.04, because higher version already has libfaudio0 in repository)
    >
    > ```shell
    > sudo add-apt-repository ppa:cybermax-dexter/sdl2-backport
    > sudo apt update
    > ```

    For [debian 10](https://computingforgeeks.com/how-to-install-wine-on-debian/):
    ```
    sudo dpkg --add-architecture i386

    wget -qO - https://dl.winehq.org/wine-builds/winehq.key | sudo apt-key add -

    sudo apt-add-repository https://dl.winehq.org/wine-builds/debian/

    wget -O- -q https://download.opensuse.org/repositories/Emulators:/Wine:/Debian/Debian_10/Release.key | sudo apt-key add -

    echo "deb http://download.opensuse.org/repositories/Emulators:/Wine:/Debian/Debian_10 ./" | sudo tee /etc/apt/sources.list.d/wine-obs.list
    ```

    update:

    ```
    sudo apt update
    ```

    Notes:

    > check linux distribution:
    > ```
    > cat /etc/*-release
    > ```
    > see: https://www.cyberciti.biz/faq/find-linux-distribution-name-version-number/

    > If you encounter an error:
    >
    > ```
    > sudo: apt-add-repository: command not found
    > ```
    >
    > install
    >
    > ```
    > sudo apt -y install software-properties-common
    > ```



3. Install latest wine

    [How to Install Wine Devel 4.8 in Ubuntu 19.04 / 18.04 | UbuntuHandbook](http://ubuntuhandbook.org/index.php/2019/05/nstall-wine-4-8-ubuntu-19-04-18-04/)

    ```shell
    sudo apt install --install-recommends winehq-devel
    ```

4. install python in wine

    [python - compiling .py into windows AND mac executables on Ubuntu - Stack Overflow](https://stackoverflow.com/questions/17709813/compiling-py-into-windows-and-mac-executables-on-ubuntu)

    ```shell
    wine --version
    winecfg

    wget https://www.python.org/ftp/python/3.6.8/python-3.6.8-amd64.exe
    wine python-3.6.8-amd64.exe
    ```

    headless install:

    > install xvfb:
    > ```
    > sudo apt-get install --fix-missing -y xvfb
    > ```
    >
    > cerate virtual frame buffer:
    > ```
    > Xvfb :0 -screen 0 1024x768x16 &
    > ```
    >
    > execute python installer:
    > ```
    > DISPLAY=:0.0 wine python-3.6.8-amd64.exe /quiet InstallAllUsers=1 PrependPath=1
    > ```
    see: [xorg - Run wine totally headless - Super User](https://superuser.com/questions/902175/run-wine-totally-headless)


5. install wxpython, gooey, pyinstaller and other requirements

    ```shell
    wine pip install wxpython
    wine pip install gooey
    wine pip install -r project_requirements.txt
    wine pip install pyinstaller==3.5
    ```

6. activate upx compression (optional):

    [Releases · upx/upx](https://github.com/upx/upx/releases)

    ```shell
    wget https://github.com/upx/upx/releases/download/v3.96/upx-3.96-win64.zip
    unzip -j "upx-3.96-win64.zip" "upx-3.96-win64/upx.exe" -d "./"
    ```

    note: upx may cause dll corruption, you may want to disable UPX, see:

    - [Onefile builds not working on Windows 10, Error loading Python dll · Issue #3600 · pyinstaller/pyinstaller](https://github.com/pyinstaller/pyinstaller/issues/3600)
    - [python - Error when creating executable file with pyinstaller - Stack Overflow](https://stackoverflow.com/questions/38811966/error-when-creating-executable-file-with-pyinstaller)
    - [python - DLL load failure with Python3 (32bit)+PyInstaller+UPX (32bit) under Windows 10 (64bit) - Stack Overflow](https://stackoverflow.com/questions/59034735/dll-load-failure-with-python3-32bitpyinstallerupx-32bit-under-windows-10)
    - [Dependency Walker (depends.exe) Home Page](https://www.dependencywalker.com/)

### Step 7: Building Windows Executable from Linux

```shell
wine pyinstaller -F --windowed build-win.spec
```

debug mode

[Using PyInstaller — PyInstaller 3.6 documentation](https://pyinstaller.readthedocs.io/en/stable/usage.html#how-to-generate)

```shell
wine pyinstaller -F --windowed build-win.spec -d bootloader
```

### Step 8: Running Windows Executable through wine

```shell
wine ./dist/DataTag_helper.exe
```

## Run test

Run test

```sh
pytest
```

Run test with print messages

```sh
pytest -s
```

Run single test function

```sh
pytest -s tests/test_DataTag_helper.py::test_json_to_excel_1
```


How to test with cli? see: [Build and Test a Command Line Interface with Poetry, Python's argparse, and pytest - DEV Community 👩‍💻👨‍💻](https://dev.to/bowmanjd/build-and-test-a-command-line-interface-with-poetry-python-s-argparse-and-pytest-4gab)

## References

### Gooey

- [chriskiehl/Gooey: Turn (almost) any Python command line program into a full GUI application with one line](https://github.com/chriskiehl/Gooey)

### Gooey Options

- [Gooey/Gooey-Options.md at master · chriskiehl/Gooey](https://github.com/chriskiehl/Gooey/blob/master/docs/Gooey-Options.md)

### examples

- [chriskiehl/GooeyExamples: Example programs to Demonstrate Gooey's functionality](https://github.com/chriskiehl/GooeyExamples)

### mutually_exclusive_group

- [Gooey/all_widgets_subparser.py at 66f154decb8cd7d140a702277d8c262c4d8194cc · chriskiehl/Gooey](https://github.com/chriskiehl/Gooey/blob/66f154decb8cd7d140a702277d8c262c4d8194cc/gooey/tests/integration/programs/all_widgets_subparser.py#L75-L80)

- [python - How to define a mutually exclusive group of two positional arguments? - Stack Overflow](https://stackoverflow.com/questions/15530117/how-to-define-a-mutually-exclusive-group-of-two-positional-arguments)

- [Python argparse mutual exclusive group - Stack Overflow](https://stackoverflow.com/questions/17909294/python-argparse-mutual-exclusive-group)

### widget example

FileChooser, DirChooser, FileSaver, MultiFileChooser, DateChooser, TimeChooser

- [GooeyExamples/widget_demo.py at d0606a8d543babe1cff5db215d4d37e26888b490 · chriskiehl/GooeyExamples](https://github.com/chriskiehl/GooeyExamples/blob/d0606a8d543babe1cff5db215d4d37e26888b490/examples/widget_demo.py#L39-L54)

### dynamic update

- [GooeyExamples/dynamic_updates.py at master · chriskiehl/GooeyExamples](https://github.com/chriskiehl/GooeyExamples/blob/master/examples/dynamic_updates.py)


CHANGELOG.md
## changelog

### v0.9.0

1. Support ArticleTag column

### v0.8.9

1. Support duplicate content by using title_hash - content_hash as TextID in excel file


### v0.8.8

1. add excel to json support

    ![excel_column](./assets/excel_column.png)

    Title 就是文章標題，
    Poster 為發文者，
    Gender 發文者為性別，
    Date 為日期(年月日)，
    Time 為時間(時分秒)，
    Content 為文章內容

    轉出來的 json 檔，會變成四個欄位：
    Title, Content, Author, Time

    Poster 跟 Gender 會合併為 Author
    Date 跟 Time 會合併為 Time

    如下所示：
    ```json
    "Title": "Irure incididunt ipsum velit anim amet occaecat elit esse cillum amet excepteur.",
    "Content": "Ex veniam ad cupidatat nostrud excepteur in ipsum. Cupidatat Lorem duis ipsum laboris sit ex sunt ea dolore cupidatat sint consectetur duis. Dolor dolore esse ullamco excepteur consectetur eiusmod sit. Irure reprehenderit pariatur et laboris. Dolor incididunt nulla voluptate magna enim duis occaecat mollit dolor Lorem enim anim pariatur.",
    "Author": "匿名/女",
    "Time": "2019-07-02/16:45:17"
    ```

### v0.8.7

1. add .docx support
2. add .pdf support
3. Support duplicate content by using title_hash - content_hash as TextID

### v0.8.6

1. exported title in json file use relative path

    ![](./assets/Snipaste_2023-02-09_10-08-42.png)

### v0.8.5

1. Set default selection to the first option, and move input dir that contains .txt to the first option.

    ![](./assets/Snipaste_2023-02-08_18-16-48.png)

2. Rename title to DataTag Helper (標註轉檔小幫手)



### v0.8.4

1. Move SqliteDict into main(), only when using GUI will call it. This is to avoid connecting to the sqlite when test CLI.

### v0.8.3

1. Rename to DataTag_helper

### v0.8.2

1. fix "Missing optional dependency 'xlrd'." error, jsut use openpyxl. see: https://github.com/pandas-dev/pandas/issues/38424

2. called monkey patch for gooey gui components only when before entering GUI mode (to avoid non-gui import generate "module not found" error)

### v0.8.1

second labeled json cleaner support input Multiple first labeled file

![choose_multiple_file](./assets/choose_multiple_file.gif)


### v0.8.0

1. add second_upload function

    ![Screenshot-of-second_upload](./assets/Deepin%20截圖_DataTag_helper.py_20210503171532.png)

2. add second labeled json cleaner and converter

    ![Screenshot-of-second_labeled](./assets/Deepin%20截圖_DataTag_helper.py_20210503171946.png)

### v0.7.2

1. add SerialID column in contents tab

    ![Screenshot-of-SerialID](./assets/Deepin%20截圖_選取範圍_20210326162722.png)


### v0.7.1

1. fix numpy issue
2. remove xlrd
3. fix KeyError issue

### v0.7.0

- 新增批次處理txt檔功能

    ![DataTag_helper_v0.7](./assets/Deepin%20截圖_DataTag_helper.py_20201214223220.png)

    注意：第一層的txt會自動忽略，只會加入第二層以下的txt

    ![folder_structure](assets/Deepin%20截圖_選取範圍_20201214234918.png)

- 已標註資料轉換輸出之 excel檔新增多種視圖：

    a. 以標註者為單位，一次呈現多位標著者之選項及句子標註結果
    ![sent_doc_cmp](./assets/Deepin%20截圖_libreoffice_20201214225512.png)

    b. 以選項為單位，比較多位標註者之結果
    ![doc_label_cmp](./assets/Deepin%20截圖_libreoffice_20201214224606.png)

    c. 以句子為單位，比較不同標註者結果
    ![sent_label_cmp](./assets/Deepin%20截圖_libreoffice_20201214224717.png)

    d. 以句子編號為單位，比較不同標註者結果
    ![sent_label_cmp(wide)](./assets/Deepin%20截圖_libreoffice_20201214224838.png)

    e. 將句子按類別及編號水平展開呈現
    ![sentence_label(wide)](./assets/Deepin%20截圖_libreoffice_20201214225322.png)

    f. 將句子按類別水平展開呈現
    ![sentence_label](./assets/Deepin%20截圖_libreoffice_20201214225227.png)

- 未標註 excel 檔轉換成json 時，輸出 TextID mapping

    原本的 TextID 是從 excel 中的 Content 全文計算md5 hash得到，
    但是因為excel中可能存在非法字元及表情符號，
    後來輸出的文章會經過移除非法字元及將表情符號轉成文字的處理，
    因此有可能無法還原成原始的全文，造成無法計算得到原始的TextID。
    所以另外計算一個TextID 是透過將全文移除非法字元及表情符號後再計算md5，
    並且輸出處理後的TextID 與原始TextID 的對應表，
    如此當文章同一篇文章經過處理後，因原始 TextID不同而對不上時，
    可透過 TextID mapping 比較處理後的TextID，
    應該就能對回同一篇文章。

    ![TextID mapping](assets/Deepin%20截圖_選取範圍_20201214232728.png)

### v0.6.1

- Symptom1: 修復 Windows 下合併檔案出現 openpyxl.utils.exceptions.IllegalCharacterError

    Rootcause: The previous generated excel file contains illegal character \_x0008\_, which is an OOXML escape character[1].

    Solution: remove these illegal characters before write to excel[2].

- Symptom2: 修復當 json 檔中有 \_xHHHH\_ 這類字串(以純文字形式出現)時，輸出的 excel 檔會自動轉換成 \_xHHHH\_ 的 unicode 字元(僅在 Windows 下發生，在 linux 下會將純文字的底線(underscore, _) 再跳脫一次，轉換成 \_x005F_x0008\_ [1] 儲存到 excel)

    Rootcause: maybe openpyxl bug?

    Solution: this is a workaround. when dataframe read from json file by read_json(), look for \_xHHHH\_ pattern in this dataframe, and unescape it[3] before write to excel.

- References

    [1]:[VTBString Class (DocumentFormat.OpenXml.VariantTypes) | Microsoft Docs](https://docs.microsoft.com/en-us/dotnet/api/documentformat.openxml.varianttypes.vtbstring?view=openxml-2.8.1)

    > [ISO/IEC 29500-1 1st Edition]
    >
    > **bstr (Basic String)**
    >
    > This element defines a binary basic string variant type, which can store any valid Unicode character. Unicode characters that cannot be directly represented in XML as defined by the XML 1.0 specification, shall be escaped using the Unicode numerical character representation escape character format \_xHHHH\_, where H represents a hexadecimal character in the character's value. [*Example*: The Unicode character 8 is not permitted in an XML 1.0 document, so it shall be escaped as \_x0008\_. *end example*] To store the literal form of an escape sequence, the initial underscore shall itself be escaped (i.e. stored as \_x005F\_). [*Example*: The string literal *\_x0008\_* would be stored as *\_x005F_x0008\_*. *end example*]
    >
    > The possible values for this element are defined by the W3C XML Schema *string* datatype.

    [2]:[(1条消息)openpyxl.utils.exceptions.IllegalCharacterError 错误原因分析及解决办法_村中少年的专栏-CSDN博客](https://blog.csdn.net/javajiawei/article/details/97147219)

    > 进入python命令行模式，输入如下：
    >
    > ```python
    > >>> import sys
    > >>> help('openpyxl')
    >
    > ```
    >

old_test/工研院標註檔案json轉xlsx.py

# coding: utf-8

# # 工研院json轉excel

# ## 讀取標註檔案

# In[1]:


import pandas as pd

# filepath = './A1_labeled.json'
filepath = '工研院標註檔案_test.json'
# filepath = '完整_二次標註_1091014.json'

df0 = pd.read_json(filepath)
df0


# ## 輸出excel

# In[2]:


import sys
sys.path.append("..")
from DataTag_helper import to_excel_AI_clerk_labeled_data

# filepath2 = './A1_labeled.xlsx'
filepath2 = './工研院標註檔案_test_labeled.xlsx'
# filepath2 = './完整_二次標註_1091014_labeled.xlsx'

df_content, df_document_label, df_sentence_label, df_sentence_label_wide, df_doc_label_cmp, df_sent_label_cmp_long, df_sent_label_cmp_wide, df_sent_doc_cmp = to_excel_AI_clerk_labeled_data(df0, filepath2)


# In[3]:


df_content


# In[4]:


df_document_label


# In[5]:


df_sentence_label


# In[6]:


df_document_label[pd.isnull(df_document_label['Crisis_Level'])]


# In[7]:


df_sentence_label[pd.isnull(df_sentence_label['Crisis_Level'])]


# In[91]:


df1 = pd.read_excel('A1.xlsx')


# In[102]:


df1[df1['Title'] == '好累卻不能說…甚至不了解自己']['Content'].iloc[0]


# In[93]:


df2 = pd.read_excel('NEW上傳_440_二次標註檔案_1090903_emojilized.xlsx')


# In[103]:


df2[df2['Title'] == '好累卻不能說…甚至不了解自己']['Content'].iloc[0]


# In[8]:


df3 = pd.read_excel('NEW上傳_440_二次標註檔案_1090903_TextID_mapping.xlsx')
df3


# In[104]:


df_A1 = pd.read_excel('A1_TextID_mapping.xlsx')


# In[105]:


df_A2 = pd.read_excel('A2_TextID_mapping.xlsx')


# In[106]:


df_all = pd.concat([df_A1, df_A2], axis=0)


# In[107]:


df_all


# In[117]:


df_sec = pd.read_excel('NEW上傳_440_二次標註檔案_1090903_TextID_mapping.xlsx')


# In[118]:


df_not_same = df_sec[~(df_sec['TextID(processed)'].isin(df_all['TextID(processed)']))]


# In[119]:


df_content[df_content['TextID'].isin(df_not_same['TextID'])]


# In[114]:


import re
pattern2 = re.compile(u'\\\\\\\\%')


# In[116]:


pattern2.sub('%', u'還不是@#$\\\\%\\\\%…\n\n容易')


# In[9]:


if all(df_content['TextID'].isin(df3['TextID'])):
    df4 = pd.merge(df3, df_content, how='right', on=['TextID'])
    display(df4)
else:
    print('false')


# In[10]:


df4_tmp = df4[~(df4['TextID(emojilized)'] == df4['TextID'])]
df4_tmp


# In[11]:


df4[~(df4['TextID(remove_illegal_characters)'] == df4['TextID'])]


# In[12]:


df4[~(df4['TextID(deemojilized)'] == df4['TextID'])]


# In[13]:


df5 = pd.read_excel('all_data.xlsx')
df5


# In[14]:


df4_tmp[~df4_tmp['TextID(emojilized)'].isin(df5['TextID'])]


# In[46]:


df1 = pd.read_excel('A2.xlsx')


# In[47]:


import hashlib



text1 = df1[df1['Title'] == '⚠️（最後一更）我被家暴了']['Content'].iloc[0]
print(hashlib.md5(text1.encode('utf-8')).hexdigest()[:10])
text1


# In[31]:


df2[df2['TextID'] == '76c492c986']


# In[48]:


df2 = pd.read_excel('NEW上傳_440_二次標註檔案_1090903_emojilized.xlsx')


# In[49]:


text2 = df2[df2['Title'] == '⚠️（最後一更）我被家暴了']['Content'].iloc[0]
print(hashlib.md5(text2.encode('utf-8')).hexdigest()[:10])
text2


# In[52]:


import difflib

case_a = 'afrykbnerskojęzyczny'
case_b = 'afrykanerskojęzycznym'

output_list = [li for li in difflib.ndiff(case_a, case_b) if li[0] != ' ']


# In[53]:


output_list


# In[66]:


[li for li in difflib.ndiff(text1, text2)][17:20]


# In[70]:


text1[17:20].encode('unicode-escape')


# In[213]:


text1[13:15].encode('unicode-escape')


# In[117]:


b'  \\u2764'.decode('unicode-escape')


# In[19]:


'◀️'.encode('unicode-escape')


# In[116]:


u'\U0001F1FF'


# In[189]:


difflib.ndiff(text1, text2)


# In[211]:


u'\uFE0F'.encode('unicode-escape')


# In[217]:


[li for li in difflib.ndiff(text1, text2)]


# In[1]:


import emoji

emoji.emojize('\n:angry_face_with_horns:\n', variant="emoji_type").encode('unicode-escape')


# In[132]:


'♀'.encode('unicode-escape')


# In[121]:


import re

with open("emoji-variation-sequences.txt", "r") as text_file:
#     print(text_file.read())
    text = text_file.read()

new_text = re.sub('#.*[\n]', '', text)
new_text2 = re.sub('(text|emoji) style;', '', new_text)
new_text3 = re.sub('\s*(FE0E\s*|FE0F\s*)', '', new_text2)
text4 = re.split(';\s*', new_text3)
newlist = list(dict.fromkeys(text4))[:-1]

# newlist

newlist2 = [('\\u' + code).encode('utf8').decode('unicode-escape') for code in newlist]
newlist2


# In[120]:


newlist2[0].decode('unicode-escape')


# In[103]:


'\u000'.decode('unicode')


# In[146]:


string = u'\U0001F926\U0000200D\U00002640'.encode('unicode-escape').decode('utf8')


# In[164]:


('\\' + string.split('\\')[-1]).encode('utf8').decode('unicode-escape')


# In[8]:


for i in newlist2:
    print('{}, '.format(i))


# In[65]:


type(newlist[0])


# In[67]:


'\\U000'


# In[42]:


with open("emoji-variation-sequences.txt", "r") as text_file:
#     print(text_file.read())
    text = text_file.read()


# In[43]:


print(text)


# In[74]:


import re

new_text = re.sub('#.*[\n]|\n*', '', text)


# In[75]:


new_text


# In[86]:


new_text2 = re.sub('\s*(text|emoji) style;\s*', '', new_text)


# In[87]:


new_text2


# In[88]:


new_text3 = re.sub('\s*(FE0E\s*|FE0F\s*)', '', new_text2)


# In[89]:


new_text3


# In[90]:


text4 = re.split(';\s*', new_text3)


# In[91]:


text4


# In[55]:


newlist = list(dict.fromkeys(text4))


# In[58]:


newlist


# In[33]:


new_text.split(';')


# In[29]:


from pandas.compat import BytesIO

memfile = BytesIO()

memfile.write(new_text.encode('utf8'))


# In[26]:


with open(memfile, "w") as text_file:
#     print(text_file.read())
    text_file.write('{}'.format(new_text))


# In[31]:


pd.read_csv(memfile)

